### PR TITLE
Milestone status

### LINKED ISSUE #1261: As a developer, I want to see what is the milestones persisted on a node
### Description
We need a notion of how far back a node is synced in `milestones`.

### Motivation
Clients need a way of determining the transaction history available on a node. 

### Requirements
1. I query a node. 
2. I get back a result which allows me to see the:
      -  oldest milestone a node persists in its database (oldest unpruned milestone) 
      - the last snapshotted milestone (the last milestone on which a local snapshot was performed)
      - the latest milestone the node persists

### Implementation
We agreed on exposing this under `getNodeInfo`, the naming is up for discussion:
- `oldestMilestoneIndex` (current `milestoneStartIndex`)
- `lastSnapshottedMilestoneIndex` (this one is new)
- `latestMilestoneIndex` (we have this already)

I think we can expose this as separate fields right under `getNodeInfo`, but please give some thoughts to how this works with the current milestone properties.

### PR DESCRIPTION
**Pull request by kwek20**
# Description
We need a notion of how far back a node is synced in milestones in order to properly query for transactions on a node.

Fixes #1261

## Type of change
- Enhancement (a non-breaking change which adds functionality)

# How Has This Been Tested?
Unit tests
getNodeInfo returns expected information

# Checklist:
- [ ] My code follows the style guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by DyrellC** [comment]
Tested, and the api responses are returning updated snapshot index milestone as expected.

File: src/main/java/com/iota/iri/service/transactionpruning/jobs/MilestonePrunerJob.java [hunk 1]
```diff
@@ -125,6 +127,7 @@ public void process() throws TransactionPruningException {
                     }
 
                     getTransactionPruner().saveState();
+                    getSnapshot().setIndex(getCurrentIndex());
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/service/transactionpruning/jobs/MilestonePrunerJob.java]: This incorrect,
The snapshot currentIndex (the milestone where we took the snapshot) shouldn't be the last pruned index...

Remember snapshotting and pruning are two separate things

File: src/main/java/com/iota/iri/service/API.java [hunk 2]
```diff
@@ -874,17 +876,33 @@ private AbstractResponse interruptAttachingToTangleStatement(){
       **/
     private AbstractResponse getNodeInfoStatement(){
         String name = instance.configuration.isTestnet() ? IRI.TESTNET_NAME : IRI.MAINNET_NAME;
-        return GetNodeInfoResponse.create(name, IRI.VERSION,
+        
+        // Note: When the MilestonePrunerQueue is processing a job, it is possible that this index is out of date.
+        // Chances of this increase when a large localSnapshotsIntervalSynced is chosen.
+        // If pruning is off, return -1
+        MilestonePrunerJobQueue job = instance.transactionPruner.getJobQueueByQueueClass(MilestonePrunerJobQueue.class);
```
[the remark this answers is not in the document]
Comment (kwek20) [reply src/main/java/com/iota/iri/service/API.java]: But then it doesnt fit on the line :(

File: src/main/java/com/iota/iri/service/transactionpruning/jobs/MilestonePrunerJob.java (hunk 1 above)
Comment (kwek20) [reply src/main/java/com/iota/iri/service/transactionpruning/jobs/MilestonePrunerJob.java]: Oh good catch, i thought i deleted that...

File: src/main/java/com/iota/iri/service/API.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (GalRogozinski) [reply src/main/java/com/iota/iri/service/API.java]: If you do the fix in the below comment than you won't need `job` :-)

File: src/main/java/com/iota/iri/service/transactionpruning/TransactionPruner.java [hunk 3]
```diff
@@ -1,5 +1,7 @@
 package com.iota.iri.service.transactionpruning;
 
+import com.iota.iri.service.transactionpruning.async.JobQueue;
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/service/transactionpruning/TransactionPruner.java]: is this import being used?

**Comment by GalRogozinski** [comment]
image


Spamming `getNodeInfo` on 4 threads with Jmeter
It seems like performance is slowly degrading than stabilizing.

Seems fine to me

**Pull request merged**